File: rerun_py/rerun_sdk/rerun/components/view_coordinates2d_ext.py
from __future__ import annotations

from enum import IntEnum
from typing import TYPE_CHECKING, Any, cast

import numpy as np
import numpy.typing as npt

if TYPE_CHECKING:
    from .._baseclasses import DescribedComponentBatch
    from ..components import ViewCoordinates2D as ViewCoordinates2DComponent
    from . import ViewCoordinates2D


class ViewCoordinates2DExt:
    """Extension for [ViewCoordinates2D][rerun.components.ViewCoordinates2D]."""

    class ViewDir2D(IntEnum):
        Up = 1
        Down = 2
        Right = 3
        Left = 4

    @staticmethod
    def coordinates__field_converter_override(data: npt.ArrayLike) -> npt.NDArray[np.uint8]:
        coordinates = np.asarray(data, dtype=np.uint8)
        if coordinates.shape != (2,):
            raise ValueError(f"ViewCoordinates2D must be a 2-element array. Got: {coordinates.shape}")
        return coordinates

    # Implement the AsComponents protocol
    def as_component_batches(self) -> list[DescribedComponentBatch]:
        from ..archetypes import ViewCoordinates2D

        return ViewCoordinates2D(cast("ViewCoordinates2DComponent", self)).as_component_batches()

    RD: ViewCoordinates2D = None  # type: ignore[assignment]
    """X=Right, Y=Down (default, image/screen convention)."""

    RU: ViewCoordinates2D = None  # type: ignore[assignment]
    """X=Right, Y=Up (math/plot convention)."""

    LD: ViewCoordinates2D = None  # type: ignore[assignment]
    """X=Left, Y=Down (horizontally mirrored image)."""

    LU: ViewCoordinates2D = None  # type: ignore[assignment]
    """X=Left, Y=Up (both axes flipped)."""

    @staticmethod
    def deferred_patch_class(cls: Any) -> None:
        cls.RD = cls([cls.ViewDir2D.Right, cls.ViewDir2D.Down])
        cls.RU = cls([cls.ViewDir2D.Right, cls.ViewDir2D.Up])
        cls.LD = cls([cls.ViewDir2D.Left, cls.ViewDir2D.Down])
        cls.LU = cls([cls.ViewDir2D.Left, cls.ViewDir2D.Up])

File: rerun_cpp/src/rerun/archetypes/view_coordinates2d_ext.cpp
#include "view_coordinates2d.hpp"

// <CODEGEN_COPY_TO_HEADER>
#include "../rerun_sdk_export.hpp"

// </CODEGEN_COPY_TO_HEADER>

// Uncomment for better auto-complete while editing the extension.
//#define EDIT_EXTENSION

namespace rerun::archetypes {

#ifdef EDIT_EXTENSION
    struct ViewCoordinates2DExt {
        uint8_t coordinates[2];
#define ViewCoordinates2D ViewCoordinates2DExt

        // <CODEGEN_COPY_TO_HEADER>

        /// Construct ViewCoordinates2D from x/y values.
        ViewCoordinates2D(uint8_t axis0, uint8_t axis1)
            : ViewCoordinates2D(rerun::components::ViewCoordinates2D(axis0, axis1)) {}

        /// X=Right, Y=Down (default, image/screen convention).
        RERUN_SDK_EXPORT static const rerun::archetypes::ViewCoordinates2D RD;

        /// X=Right, Y=Up (math/plot convention).
        RERUN_SDK_EXPORT static const rerun::archetypes::ViewCoordinates2D RU;

        /// X=Left, Y=Down (horizontally mirrored image).
        RERUN_SDK_EXPORT static const rerun::archetypes::ViewCoordinates2D LD;

        /// X=Left, Y=Up (both axes flipped).
        RERUN_SDK_EXPORT static const rerun::archetypes::ViewCoordinates2D LU;

        // </CODEGEN_COPY_TO_HEADER>
    };
#endif

    const ViewCoordinates2D ViewCoordinates2D::RD =
        ViewCoordinates2D(rerun::components::ViewCoordinates2D::RD);
    const ViewCoordinates2D ViewCoordinates2D::RU =
        ViewCoordinates2D(rerun::components::ViewCoordinates2D::RU);
    const ViewCoordinates2D ViewCoordinates2D::LD =
        ViewCoordinates2D(rerun::components::ViewCoordinates2D::LD);
    const ViewCoordinates2D ViewCoordinates2D::LU =
        ViewCoordinates2D(rerun::components::ViewCoordinates2D::LU);

} // namespace rerun::archetypes
